seriesfi/series-swe-interview
# Refactor/typescript gilded rosa kata

### Goal
This pull request includes a requested refactor to the `gilded-rose.ts` class and a bug fix for `Conjured` Items.

### Issue
(Since no ticket exists for this issue) The details of the issue can be found here : https://github.com/seriesfi/series-swe-interview#gilded-rose-requirements-specification.

### Steps to test
1. Checkout branch
2. Run `npm run i` to install dependencies (more details can be found in the `README.md`
3. To run the node application, run `npx ts-node test/golden-master-text-test.ts [numberOfDays]`. 
Example: `npx ts-node test/golden-master-text-test.ts 10`
4. Verify the printed output makes sense and aligns with the rules specified by Gilded Rose found here: https://github.com/seriesfi/series-swe-interview#gilded-rose-requirements-specification 
5. Verify conjured item `Conjured Mana Cake` works as intended (it was previously broken)
6. Run the tests and verify all tests pass
To run JEST, `npm run test:jest`
To run MOCHA, `npm run test:mocha` 

**Current coverage report**
```
 PASS  test/jest/gilded-rose.spec.ts
  Gilded Rose
    ✓ should foo (1 ms)
  update quality base item
    ✓ quality is decremented if conditions are met, sellin is decremented (1 ms)
  update quality aged brie
    ✓ quality is incremented, sell in is decremented (3 ms)
    ✓ quality is incremented and quality never exceeds max value
  update quality sulfaras
    ✓ quality and sellin is never incremented for sulfaras (1 ms)
  update quality backstage passes
    ✓ quality increments by 2 if sellin <= 10
    ✓ quality increments but never exceeds 50
    ✓ quality increments but drops to zero after the concert (1 ms)
  update quality conjured
    ✓ quality decreases twice as fast as normal items

-------------------|---------|----------|---------|---------|-------------------
File               | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
-------------------|---------|----------|---------|---------|-------------------
All files          |     100 |    97.91 |     100 |     100 |                   
 classes           |     100 |    93.75 |     100 |     100 |                   
  gilded-rose.ts   |     100 |    91.66 |     100 |     100 | 15                
  item.ts          |     100 |      100 |     100 |     100 |                   
 helpers           |     100 |      100 |     100 |     100 |                   
  enum.ts          |     100 |      100 |     100 |     100 |                   
  updateQuality.ts |     100 |      100 |     100 |     100 |                   
-------------------|---------|----------|---------|---------|-------------------
Test Suites: 1 passed, 1 total
Tests:       9 passed, 9 total
Snapshots:   0 total
Time:        0.472 s, estimated 1 s
Ran all test suites.
```
